Refactor the LIFX discovery process to resolve two issues

Djelibeybi opened the pull request:
Signed-off-by: Avi Miller <me@dje.li>

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added logic to ignore the LIFX Switch as the aiolifx library does
not support it. Also adjusted the discovery flow to allow
sufficient time for the async discovery to complete.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #57195 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Djelibeybi committed:
Refactor the LIFX discovery process to resolve two issues

Added logic to ignore the LIFX Switch as the aiolifx library does
not support it. Also adjusted the discovery flow to allow
sufficient time for the async discovery to complete.

Resolves #57195.

Signed-off-by: Avi Miller <me@dje.li>

Djelibeybi commented:
I've been testing this pretty extensively for the past few days and I've received zero errors from discovery both during HA startup and normal operations.

Strit commented:
That's awesome. My bulbs have been missing from my instance for the last few weeks, nothing I have done have made them come back. They do have IP's according to the dhcp server.

Djelibeybi commented:
@Strit if you've got HACS installed, you can test the beta by adding this repo: https://github.com/djelibeybi/ha-lifx-beta

Adding `custom_components.lifx: debug` to your `logger` configuration will allow you to see the registration process for each bulb, from `NEW` to `READY`.

Strit commented:
I don't have HACS. I use a docker container.

Djelibeybi commented:
> I don't have HACS. I use a docker container.

Ah. I tried to find some docs on installing custom components without HACS, but didn't find anything useful. Essentially, you just copy the `custom_components` folder into your `config` location alongside your `configuration.yaml` and it should pick it up.

Djelibeybi committed:
fix: check for an entity_id before updating

This resolves a rare NoEntitySpecifiedError when a bulb is
rediscovered but hasn't finished instantiating before the
event loop tries to update state.

Signed-off-by: Avi Miller <me@dje.li>

mekenthompson commented:
@Djelibeybi - I've got this setup and tested, seems to be working great so far. No startup delays from switches.

Djelibeybi commented:
Extended testing shows that while this change fixes some things it doesn't really address the underlying issues, so I'm closing this PR while I continue to work on other options.
